File: java/datastructure/linkedlist/leetcode/PalindromeLinkedList.java
package datastructure.linkedlist.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * 回文链表
 */
public class PalindromeLinkedList {

    /**
     * 使用一个数组存储
     */
    public boolean isPalindrome1(ListNode head) {
        List<Integer> list = new ArrayList<>();
        while (head != null) {
            list.add(head.val);
            head = head.next;
        }
        
        if (list.size() > 1) {
            for (int i = 0, j = list.size() - 1; i < j; i++, j--) {
                 if (!list.get(i).equals(list.get(j))) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * 原地遍历
     */
    public boolean isPalindrome2(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

        //先找到中间节点
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        //反转从链表头到中间的节点
        ListNode prev = null, cur = head;
        while (cur != slow) {
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }

        //从中间依次往两边遍历比较
        ListNode q = slow.next;
        if (fast == null) {
            q = slow;
        }

        while (prev != null && q != null) {
            if (prev.val != q.val) {
                return false;
            }

            prev = prev.next;
            q = q.next;
        }

        return true;
    }

    private static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }
}

File: cpp/linkedlist/palindrome_linked_list.cpp
// 回文链表

#include "vector"
using namespace std;

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    // 放入数组再比较
    bool isPalindrome1(ListNode* head) {
        vector<int> arr;
        while (head) {
            arr.push_back(head->val);
            head = head->next;
        }
        for (int i = 0, j = arr.size() - 1; i < j; i++, j--) {
            if (arr[i] != arr[j]) {
                return false;
            }
        }
        return true;
    }

    // 找到中间节点再反转再比较
    bool isPalindrome2(ListNode* head) {
        // 找到中间节点
        ListNode* slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        // 反转
        ListNode* prev = nullptr, *p = head;
        while (p != slow) {
            ListNode* next = p->next;
            p->next = prev;
            prev = p;
            p = next;
        }

        ListNode* q = slow;
        if (fast != nullptr) {
            q = slow->next;
        }

        while (prev && q) {
            if (prev->val != q->val) {
                return false;
            }
            prev = prev->next;
            q = q->next;
        }
        return true;
    }
};
